hummingbot/wallet/ethereum/watcher/websocket_watcher.py

```diff
@@ -0,0 +1,169 @@
+            self._fetch_new_blocks_task = None
+            self._network_on = False
+
+    async def connect(self):
+        try:
+            self._client = await websockets.connect(uri=self._websocket_url)
+            return self._client
+        except Exception as e:
+            self.logger().network(f"ERROR in connection: {e}")
+
+    async def disconnect(self):
+        try:
+            await self._client.close()
+        except Exception as e:
+            self.logger().network(f"ERROR in disconnection: {e}")
+
+    async def _send(self, emit_data) -> int:
+        self._nonce += 1
+        emit_data["id"] = self._nonce
+        await self._client.send(ujson.dumps(emit_data))
+        return self._nonce
+
+    async def subscribe(self, params) -> bool:
+        emit_data = {
+            "method": "eth_subscribe",
+            "params": params
+        }
+        nonce = await self._send(emit_data)
+        raw_message = await self._client.recv()
+        if raw_message is not None:
+            resp = ujson.loads(raw_message)
+            if resp.get("id", None) == nonce:
+                self._node_address = resp.get("result")
+                return True
+        return False
+
+    async def unsubscribe(self, params) -> bool:
+        emit_data = {
+            "method": "eth_unsubscribe",
+            "params": params
+        }
+        nonce = await self._send(emit_data)
+        raw_message = await self._client.recv()
+        resp = ujson.loads(raw_message)
+        if resp.get("id", None) == nonce:
+            result = resp.get("result", None)
+            return (result is not None) and result
+
+    async def fetch_new_blocks_loop(self):
+        try:
+            while True:
+                try:
+                    raw_message = await self._client.recv()
+                    message_json = ujson.loads(raw_message) if raw_message is not None else None
+                    if message_json.get("method", None) == "eth_subscription":
+                        subscription_result_params = message_json.get("params", None)
+                        incoming_block = subscription_result_params.get("result", None) \
+                            if subscription_result_params is not None else None
+                        if incoming_block is not None:
+                            new_block: AttributeDict = await self.call_async(self._w3.eth.getBlock,
+                                                                             incoming_block.get("hash"), True)
+                            self._current_block_number = new_block.get("number")
+                            self._block_cache[new_block.get("hash")] = new_block
+                            self.trigger_event(NewBlocksWatcherEvent.NewBlocks, [new_block])
+                except asyncio.CancelledError:
+                    if self._network_on:
+                        self.logger().network("Network closed abnormally, reconnecting ...")
+                        self.connect()
```

Comment:
Not sure if we should `connect()` here on this exception, seems to me the task is cancelled and exit from here. To connect here might be unnecessary.

Reply:
The Websocket sometimes "abnormally" disconnects, for example when my computer sleeps with the bot running; this accounts for the "abnormal disconnections" when there's a CancelledError but the WebSocket is still intended to be connected; this replaces the self._client with a connected WebSocket. Let me know if I should figure out a workaround or just remove it

Reply:
Does the `CancelledError` raise on those situations, like when the websocket abnormally disconnects? From what I understand the CancelledError is raised only when `cancel()` is called, and it is a way to wrap up and finish the task. If you need to reconnect, it's fine but I don't think this is the right place for it.

Reply:
Made a mistake. Wrong location; I put it in ConnectionClosedError from WebSockets instead. This is for when connectionclosed errors codes are not 1000 (OK) or 1001 ("Going Away"). Trying to figure out if there's a way to check connection another way in the fetch loop

Reply:
@Nullably pushed up changes, what do you think?